proccontrol/src/linux.C

```diff
@@ -2830,6 +2830,22 @@ bool linux_thread::plat_setRegisterAsync(Dyninst::MachRegister reg,
    return true;
 }
 
+void linux_thread::plat_handle_ghost_thread() {
+	std::string loc = "/proc/" + std::to_string(proc()->getPid()) + "/task/" + std::to_string(getLWP());
+	struct stat dummy;
+	int res = stat(loc.c_str(), &dummy);
+
+	pthrd_printf("GHOST_THREAD: exists=%d, loc=%s\n", res, loc.c_str());
+
+	if(res == -1) {
+		EventLWPDestroy::ptr lwp_ev = EventLWPDestroy::ptr(new EventLWPDestroy(EventType::Post));
+		lwp_ev->setSyncType(Event::async);
+		lwp_ev->setThread(thread());
+		dynamic_cast<linux_process*>(proc()->llproc())->decodeTdbLWPExit(lwp_ev);
```

Comment:
When running, I am seeing the error

    test_driver: /home/tim/workspace/spack/opt/spack/linux-linuxmint18-piledriver/gcc-8.3.0/boost-1.70.0-6hhwr4jmmv5a52i3wqavu6kc4xnslczo/include/boost/smart_ptr/shared_ptr.hpp:734: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = const Dyninst::ProcControlAPI::Process; typename boost::detail::sp_member_access<T>::type = const Dyninst::ProcControlAPI::Process*]: Assertion `px != 0' failed.

I believe it's this line that is causing it. I'm not sure if it's `proc` or `llproc` that is the underlying issue.

Reply:
I'm not immediately sure why this line would crash--the process containing a linux_thread is almost certainly a linux_process.  

It could be simplified to just call llproc() instead of proc()->llproc().  int_thread has a routine that returns its int_process.  But that shouldn't make a behavior difference.

Can you get in there with a debugger and see more precisely  what exactly is failing?

Could you get in there with a debugger

Reply:
>It could be simplified to just call llproc() instead of proc()->llproc().

Ah, ok. I was poking around in the types to get it to compile.

>Can you get in there with a debugger and see more precisely what exactly is failing?

That was my next step. I'll let you know what I find.

Reply:
I've finally found a reproducer, but gdb can't create a stacktrace because `test_driver` has exited normally.

@sashanicolas  @mxz297 Is there a way to run the test directly under gdb with a particular threading model (I need the MPMT version as it's the only one that consistently crashes).

    $ gdb --args ./runTests -64 -g++ -none -create -dynamiclink -nonpic -test pc_fork -log tmp.log -all

    >set env ...
    >run

    Starting program: /home/tim/paradyn/kcr4_5A6/testsuite/tests/runTests -64 -g++ -none -create -dynamiclink -nonpic -test pc_fork -log tmp.log -all
    [Detaching after fork from child process 3345]
    [Detaching after fork from child process 3347]
    [Detaching after fork from child process 3354]
    pc_fork                    g++    none 64  create   SPST    dynamic nonPIC  PASSED
    pc_fork                    g++    none 64  create   MPST    dynamic nonPIC  PASSED
    test_driver: /home/tim/workspace/spack/opt/spack/linux-linuxmint18-piledriver/gcc-8.3.0/boost-1.70.0-6hhwr4jmmv5a52i3wqavu6kc4xnslczo/include/boost/smart_ptr/shared_ptr.hpp:734: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = const Dyninst::ProcControlAPI::Process; typename boost::detail::sp_member_access<T>::type = const Dyninst::ProcControlAPI::Process*]: Assertion `px != 0' failed.
    *** Child terminated abnormally via signal 6.
    [Detaching after fork from child process 3447]
    pc_fork                    g++    none 64  create   SPMT    dynamic nonPIC  CRASHED (Running Test)
    test_driver: /home/tim/workspace/spack/opt/spack/linux-linuxmint18-piledriver/gcc-8.3.0/boost-1.70.0-6hhwr4jmmv5a52i3wqavu6kc4xnslczo/include/boost/smart_ptr/shared_ptr.hpp:734: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = const Dyninst::ProcControlAPI::Process; typename boost::detail::sp_member_access<T>::type = const Dyninst::ProcControlAPI::Process*]: Assertion `px != 0' failed.
    Child received unexpected signal: 5
    *** Child terminated abnormally via signal 6.
    [Detaching after fork from child process 3620]
    pc_fork                    g++    none 64  create   MPMT    dynamic nonPIC  CRASHED (Running Test)
    [Inferior 1 (process 3341) exited normally]
    (gdb) bt
    No stack.

Reply:
@hainest You can just replace `runTests` with `test_driver` in your gdb command line.

Reply:
For some reason, the file uploader here isn't working, so I've put the log on my google drive (https://drive.google.com/open?id=1KAUjjA_uZm1rLZOeCiOuawX4g9pBFvax).

When we pull the message out of the [mailbox](https://github.com/dyninst/dyninst/blob/master/proccontrol/src/process.C#L1096), the shared pointer has been either fully destroyed or never assigned a value.

Reply:
I stepped through the messages, and the bad message is

    Thread 2 "test_driver" hit Breakpoint 1, int_process::waitAndHandleEvents (block=<optimized out>)
        at /home/tim/paradyn/kcr4_5A6/dyninst/src/proccontrol/src/process.C:1057
    1057        Event::ptr ev = mbox()->dequeue(should_block);
    (gdb) p ev
    $45 = {px = 0x0, pn = {pi_ = 0x7ffff339ce42 <__GI___pthread_mutex_lock+258>}}

What kind of message has a pthread mutex as its payload?